### ISSUE TITLE
Should we officially "deprecate" Invoke-Expression?

### ISSUE THREAD

**Issue by sdwheeler** [PowerShell/PowerShell#11662]
# Summary of the new feature/enhancement

Opening issue to get feedback per our shiproom conversation.

We have publicly stated that `Invoke-Expression` should be avoided. There are too many inherent security risks built into it. We continue to get questions in the Docs repo about how and why it should be used, and asks for useful examples.

We have the ability, in code, to set an attribute on the cmdlet that marks it as deprecated. Should this change be made? Are there compelling cases for the use of this cmdlet?

**Comment by mklement0** [comment]
In the linked issue @sdwheeler states:

> Invoke-Expression is maintained for historical purposes but is effectively deprecated

From @rjmholt's response ibidem:

> My personal feeling (worth also talking to @SteveL-MSFT, @joeyaiello, @JamesWTruher and @BrucePay) is that it’s not so much historical as an important part of an interpreted language. It’s dangerous, and shouldn’t be used in production, but all scripting languages have an eval() equivalent.

Also of interest: his comment with specific invocation scenarios, https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4897#issuecomment-548141117, where he concludes:

> I think the uses are quite obscure and Invoke-Expression needs no encouragement. We should just warn against it..

**Comment by lukeb1961** [comment]
As someone who does use Invoke-Expression for good reason, in a small number of scripts, I'm damned if I can see why you want to take away this functionality. It exists in Powershell and exists in some form in every other scripting language.  There are plenty of compatibility  and performance issues with PS core that need to be addressed before you start breaking my existing scripts.

**Comment by doctordns** [comment]
I feel deprecating like this is not all that helpful. There are, as @lukeb1961 and others point out, useful reasons for using IEX. Not many, but some are certainly 'valid'. 

I feel that to document the issue and warn against it is probably enough.

**Comment by danijeljw** [comment]
I do not see the use in removing this, when I use it in so many scripts, and have been taught how to use this effectively in so many scripts for API calls too. Please don't remove it.

**Comment by sdwheeler** [comment]
To be clear, we are not talking about removing `Invoke-Expression`. We are talking about marking it as "deprecated" and more emphatically warning against its use.

If you have compelling use cases, please share them so we can see WHY it is needed.

**Comment by rjmholt** [comment]
> I'm damned if I can see why you want to take away this functionality

Nobody is talking about removing `Invoke-Expression`. It’s an important primitive in PowerShell (as in any scripting language).

Here and in a few other places there has been some confusion about the meaning of the term `deprecate`. It simply means to *disrecommend*. Obsolescence is one reason to do so, but another is unnecessary danger. `Invoke-Expression` lies in the second category; it’s not going anywhere but it’s [dangerous](https://devblogs.microsoft.com/powershell/invoke-expression-considered-harmful/) to use non-interactively, both from a [security standpoint](https://community.idera.com/database-tools/powershell/powertips/b/tips/posts/why-invoke-expression-is-evil) and an efficiency one (the whole parser/compiler toolchain have to be run again at execution time).

What we mean by deprecation in this case is still not clear to me (just in documentation, or with code attributes as well?), but you’ll always be able to use `Invoke-Expression` (even though you shouldn’t in scripts).

> As someone who does use Invoke-Expression for good reason, in a small number of scripts

Would you be able to share those scenarios/usages or adaptations of them? I think @sdwheeler in particular is looking for examples of `iex` usage.

**Comment by sdwheeler** [comment]
@rjmholt Thanks. Well said. The purpose of this issue was to discuss whether we should add the "deprecated" attribute to the code to enhance the warning about use of the cmdlet. This is separate from the question of how should this be documented.

I am looking for good example for how and why you are using IEX instead of the the &-operator. If someone can provide a useful example then we can add that to the documentation. Otherwise, I will just update the documentation with a more direct recommendation for not using the cmdlet.

**Comment by SeeminglyScience** [comment]
> Here and in a few other places there has been some confusion about the meaning of the term `deprecate`. It simply means to _express disapproval of_, to disrecommend.

Yeah literally every time the term is used.  Y'all may want to adopt a template for using the term in a public space, something like:

> **IMPORTANT NOTE:** *This is not about **removing** ___. ([Link to doc on what deprecation means]())*
>
> The actual post content.

Actually I don't know if anyone would read that.  Maybe a different term with less connotation needs to be used instead.

> @rjmholt Thanks. Well said. The purpose of this issue was to discuss whether we should add the "deprecated" attribute to the code to enhance the warning about use of the cmdlet. This is separate from the question of how should this be documented.

Do you mean `ObsoleteAttribute`? If so, the runtime warnings would get *really* annoying.  A PSSA rule is where this would need to live imo.

**Comment by mklement0** [comment]
>  whether we should add the "deprecated" attribute to the code to enhance the warning about use of the cmdlet. This is separate from the question of how should this be documented.

Thanks for that clarification, @sdwheeler; the distinction wasn't obvious to me, and it seems that these aspects have already been conflated here, which is just as well, because I think discussing them in tandem makes sense.

---

As for the _attribute_: I agree with @SeeminglyScience  that a PSSA rule is the right place to warn programmatically, if desired; certainly not at runtime, as that amounts to a quasi-breaking change.

---

As for _documentation_ and use cases:

In the `Invoke-Expression` help topic, I think a general warning is sufficient, as @rjmholt proposes, along the following lines (adapted from the original https://github.com/MicrosoftDocs/PowerShell-Docs/issues/2679 issue):

---

> As a matter of habit, try to avoid use of `Invoke-Expression`, because there are usually more secure and robust alternatives.

> If you do find that `Invoke-Expression` is your only choice, _carefully consider the security implications_: if a string from an (untrusted) outside source (e.g., user input) is passed directly to `Invoke-Expression`, _arbitrary commands may be executed_.

---

I think that is sufficient for this particular help topic, but _separately_, perhaps in a new `about_Invoke-Expression` conceptual help topic, we could focus on when _not_ to use `Invoke-Expression`.

Specifically, we could take widespread current and historical (mis)uses of `Invoke-Expression` and show _superior, safer alternatives_.

If we do end up agreeing on legitimate _public_ uses worth advertising (other than during PowerShell tests, as described by @rjmholt in https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4897#issuecomment-548141117), we can add them there as well.

Even in the perhaps most seductive use case - direct execution of script text downloaded from the web (e.g,  `iwr https://aka.ms/install-powershell.ps1 | iex`) - `Invoke-Expression` is ultimately ill-advised, and enhancing `Invoke-Command` to operate on `Invoke-WebRequest` output should be favored - see https://github.com/PowerShell/PowerShell/issues/5909#issuecomment-461192202

**Comment by rjmholt** [comment]
In fact [PSSA has had a rule to disrecommend `Invoke-Expression`](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/RuleDocumentation/AvoidUsingInvokeExpression.md) for a few years now

**Comment by rjmholt** [comment]
> certainly not at runtime, as that amounts to a quasi-breaking change

While I agree that adding an `ObsoleteAttribute` would be annoying and probably not what we want here, I strongly disagree that it would be a breaking change.

If adding `ObsoleteAttribute` to anything is a breaking change, then that attribute is pointless. The whole idea of attributes is that they behave as additive metadata.

Anyway, that’s a tangent.

**Comment by rjmholt** [comment]
> Specifically, we could take widespread current and historical (mis)uses of Invoke-Expression and show superior, safer alternatives.

Absolutely agreed.

I’d like to build a library of `Invoke-Expression` use-cases and document alternatives.

Unfortunately in the three issues we’ve been discussing this, we haven’t managed to record any of those uses. Hopefully we can collect some here so we can write or enhance documentation around them.

**Comment by mklement0** [comment]
To close the tangent: I called it a _quasi_-breaking change in that it would be confusing if existing code suddenly triggered warnings that weren't there before (this is what happened to the `Send-MailMessage` cmdlet, for instance).

**Comment by rjmholt** [comment]
> To close the tangent

Unfortunately that just extends the tangent (I also had Send-MailMessage in mind).

It might be best for us to open a new issue to discuss how we implement deprecation in PowerShell.

**Comment by SteveL-MSFT** [comment]
There have been only two instances where I had to use `Invoke-Expression`.  One is the short one-liner to install PowerShell:

```powershell
iex "& { $(irm https://aka.ms/install-powershell.ps1) } -UseMSI -Preview"
```

The other is in negative Pester test cases where I need the test script to be parsed correctly, but want the test case to fail:

```powershell
            try {
                $ErrorView = 'ConciseView'
                Invoke-Expression '1/d'
            }
            catch {
                $e = $_
            }
            finally {
                $ErrorView = $oldErrorView
            }
```

Nether of these are mainstream cases.  The latter could be converted to:

```powershell
                [scriptblock]::Create('1/d').Invoke()
```

**Comment by mklement0** [comment]
@SteveL-MSFT, the first case is the one I wish we could eliminate by the suggested [enhancement to `Invoke-Command`](https://github.com/PowerShell/PowerShell/issues/5909#issuecomment-461192202), because the `iex` solution is (a) awkward (need to use embedded `& { ... }` to avoid running directly in the caller's scope), but, more importantly, (b) _brittle_, because the presence of an `exit` statement in the downloaded script content would exit the PowerShell process as a whole.

**Comment by mklement0** [comment]
Good point about the existing PSSA rule, @rjmholt: I just realized that you can see it in action with `Invoke-ScriptAnalyzer -ScriptDefinition 'Invoke-Expression "1+1"'`, but it's not enabled by default in Visual Studio Code - I've opened https://github.com/PowerShell/PowerShellEditorServices/issues/1166 to suggest changing that.

**Comment by SeeminglyScience** [comment]
> While I agree that adding an `ObsoleteAttribute` would be annoying and probably not what we want here, I strongly disagree that it would be a breaking change.
> 
> If adding `ObsoleteAttribute` to anything is a breaking change, then that attribute is pointless. 

It definitely *can* break things (I know there is a difference between something breaking the change contract and a change technically breaking a thing, but bear with me):

```powershell
function Test-Something {
    [CmdletBinding()]
    param()
    end {
        if ($theWorldIsEnding) {
            $PSCmdlet.WriteWarning('some bad stuff is happening')
        }
    }
}
```

This is part of why I don't really like the warning system at all.  Most of the things I would want to warn about are better off as a non-terminating error anyway.

If something is writing a warning, I'm either silencing it, or throwing on it.  If I'm throwing on it with `-WarningAction Stop`, then adding `ObsoleteAttribute` will cause my script to always throw.

I know that if it's considered a breaking change than that makes the attribute useless, but that's how it feels currently. Personally I think `ObsoleteAttribute` should be moved to a design time warning only, *maybe* with `error: true` being the exception.  That's how C# works as well, anything already built keeps on chugging but you're gonna know about it if you try to use it in something new.

> The whole idea of attributes is that they behave as additive metadata.

That's true up until the point where something reads that metadata and changes behavior based on it. Adding `Mandatory` to a `ParameterAttribute` for example is for sure a breaking change.

**Comment by mklement0** [comment]
Such a strapping young tangent deserves a proper home: #11674

@SeeminglyScience, I've linked to your previous comment there, but perhaps you want to post your concerns there directly.

**Comment by xtqqczze** [comment]
`Invoke-Expression` is used extensively in our build module. #9459 is stale PR to remove.

**Comment by JrrRomeroJose** [comment]
Hello i was looking for best practices when running cmd on powershell and somehow reached this issue.
here i let you my example of use for invoke-Expression. I'd like to hear any different approach.



```Powershell
    $command = "C:\""Program Files""\MongoDB\Server\4.2\bin\mongoimport.exe --db=$($db) --collection=$($Collection) --file=$($_.FullName)"
    Write-Host "Importing file $($_.Name)" -ForegroundColor Green
    Invoke-Expression $command
```

Thanks.

**Comment by mklement0** [comment]
Always invoke external programs directly; if they're given with a quoted path / path in a variable, prepend `&`, the [call (execute) operator](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Operators#call-operator-). If you need to define a command for _later_ execution, put it in a script block (`{ ...}`) and call it on demand with `&`

```
& "C:\Program Files\MongoDB\Server\4.2\bin\mongoimport.exe"--db=$($db) --collection=$($Collection) --file=$($_.FullName)
```

For more information see [this SO answer](https://stackoverflow.com/a/57966347/45375)

**Comment by rjmholt** [comment]
> Invoke-Expression is used extensively in our build module. #9459 is stale PR to remove.

Hmmmm. It looks like all those invocations could be replaced with a `&` and passing the arguments as literals. But the `[scriptblock]::Create()` method does exactly the same thing as `Invoke-Expression`. Perhaps we should write that into the rule?

@JrrRomeroJose In your case @mklement0 is spot on. One point that comes up there sometimes is that `Invoke-Expression` on a string makes it easier to record exactly what the command executed was. However, that's exactly what [transcription](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.host/start-transcript?view=powershell-7) is intended to do.

**Comment by fedorbirjukov** [comment]
I am using Invoke-Expression to stop on non-zero exit code because $ErrorActionPreference = 'Stop' does not stop in that case.

Helper function
```
function Invoke-ExpressionWithExitCode($Command) {
    Invoke-Expression "& $Command"
    if ($LastExitCode) {
        throw "Error: non-zero exit code ($($LastExitCode)) for command: $Command"
    }
}
New-Alias -Name iex0 -Value Invoke-ExpressionWithExitCode
```

Example (execution stops only after the second clone)
```
PS> $ErrorActionPreference = 'Stop'
PS> git clone wrongurl; iex0 "git clone wrongurl"; git clone wrongurl
fatal: repository 'wrongurl' does not exist
fatal: repository 'wrongurl' does not exist
Error: non-zero exit code (128) for command: git clone wrongurl
At line:4 char:9
+         throw "Error: non-zero exit code ($($LastExitCode)) for comma ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Error: non-zero... clone wrongurl:String) [], RuntimeException
    + FullyQualifiedErrorId : Error: non-zero exit code (128) for command: git clone wrongurl
```

**Comment by mklement0** [comment]
@fedorbirjukov, here's how to do it without  `Invoke-Expression`, which has the added advantage of being able to use regular invocation syntax and avoiding headaches related to embedded quoting.

```powershell
function Invoke-Utility {
  $exe, $argsForExe = $Args
  $ErrorActionPreference = 'Stop' # in case $exe isn't found
  & $exe $argsForExe
  if ($LASTEXITCODE) { Throw "$exe indicated failure (exit code $LASTEXITCODE; full command: $Args)." }
}
Set-Alias iu Invoke-Utility

# Invoke as you would normally, just prepend "iu "
iu git clone wrongurl
```

**Comment by fedorbirjukov** [comment]
@mklement0 
Thanks for replying. I've seen this snippet before. But only now am I seeing why iex0 is worse than iu. Mostly because it does more than one call... I even uncovered a bug in my script thanks to this.

**Comment by rjmholt** [comment]
PowerShell's own build actually defines a function for this too:

https://github.com/PowerShell/PowerShell/blob/aefb82a3087a6e2fec859fe657337b1a9985e886/test/tools/Modules/HelpersCommon/HelpersCommon.psm1#L267-L312

**Comment by mklement0** [comment]
@fedorbirjukov, I've had to update the `Invoke-UtilityFunction` above to work around the bug described in #4002, which causes `2>` stderr redirections while `$ErrorActionPreference = 'Stop'` is in effect to trigger script-terminating errors; with the update, you can now call `iu git clone 2>$null` to silence stderr output without ill effects (while still enforcing a nonzero exit code).

Tangent alert:

* All current problems with `2>` stderr redirections are summarized in [this SO answer](https://stackoverflow.com/a/60115289/45375).
* A _synchronous_ function such as `Start-NativeExecution` shouldn't use the `Start` verb, which is for _asynchronous_ operations (the `Start-Sleep` cmdlet, which sleeps _synchronously_ is an unfortunate, very public exception).
* One shouldn't have to write wrapper functions at all, given that calling external (native) programs with proper error handling should be a shell's core mandate - see [this RFC draft](https://github.com/PowerShell/PowerShell-RFC/pull/88/files).
* For a wrapper function that compensates for PowerShell's broken handling of quoting of embedded `"` chars. / empty arguments when calling external programs, see https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-562334606
</sup>

**Comment by yecril71pl** [comment]
> Here and in a few other places there has been some confusion about the meaning of the term `deprecate`. It simply means to _express disapproval of_, to disrecommend. Obsolescence is one reason to do so, but another is unnecessary danger. `Invoke-Expression` lies in the second category; it’s not going anywhere but it’s [dangerous](https://devblogs.microsoft.com/powershell/invoke-expression-considered-harmful/) to use non-interactively, both from a [security standpoint](https://community.idera.com/database-tools/powershell/powertips/b/tips/posts/why-invoke-expression-is-evil) and an efficiency one (the whole parser/compiler toolchain has to be run again at execution time).

If `Invoke-Expression` gets deprecated, there is an easy workaround: put the script into a temporary file and execute that one.  Write a function that does is for you and you are all set.  Deprecating does not make any sense to me.

**Comment by rkeithhill** [comment]
Say, you marked the command with `[Obsolete("This command is dangerous in non-interactive scenarios and its use is not recommended")]`.  Where would we see that?  Personally, I do not want to see that at runtime.  I suppose the help system could detect and show this attribute:
```powershell
PS> help Invoke-Expression


OBSOLETE: This command has been marked as obsolete. This command is dangerous in non-interactive scenarios 
and its use is not recommended

NAME
    Invoke-Expression

SYNOPSIS
    Runs commands or expressions on the local computer.

...
```
And I suppose PSSA could detect this attribute and use it to flag usage of obsolete commands.  Is there anywhere else this would be surfaced?

**Comment by JrrRomeroJose** [comment]
Maybe a one time Warning. Just like Windows Powershell has his own message 'Try the new cross-platform....' but instead of being there always you should only see it one time. moreover, You could add a link to a good post that explain why using it and why not and the workaround that you have.

![imagen](https://user-images.githubusercontent.com/42807262/87651092-e5827800-c752-11ea-9873-5a1a0cec81b4.png)

**Comment by vexx32** [comment]
I don't think a warning in the shell is particularly useful. Quite possibly a lot of folks may end up seeing it from code that isn't even theirs and have no real recourse, so completely ignore it.

A PSSA rule is a much more effective strategy in my mind.

**Comment by mklement0** [comment]
Given the length of this thread, let me try to summarize:

Using `[Obsolete()]` would indeed currently surface the warning _at runtime_ (as you currently get with `1 | Format-Hex -Raw`, for instance), which I agree is undesirable.
   * The general deprecation discussion in #11674 also touches on whether even uses of `[ObsoleteAttribute()]` should ever surface _at runtime_, or whether they should surface at runtime only _on demand_ and/or possibly by default only in direct, interactive calls.
   * Surfacing them categorically (also) via PSSA is also suggested there (but there are limits to recognizing the use of deprecated features through static analysis alone.)

Assuming that there's agreement that runtime warnings should be avoided at least in this case (which is what it sounds like), the pieces to the puzzle are mostly already in place:

 * A PSSA rule for `Invoke-Expression` (specifically, with no `ObsoleteAttribute` involved) [_already exists_](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/RuleDocumentation/AvoidUsingInvokeExpression.md), it just needs to be [_turned on by default_](https://github.com/PowerShell/PowerShellEditorServices/issues/1166) (there has been no activity since this suggestion was posted on 23 January 2020).

 * `Get-Help Invoke-Expression` already [shows a warning](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7.1), which in the console (somewhat awkwardly) surfaces as follows:

```none
 > [!CAUTION] > Take reasonable precautions when using the `Invoke-Expression` cmdlet in scripts. When using >
`Invoke-Expression` to run a command that the user enters, verify that the command is safe to run > before
running it. In general, it is best to design your script with predefined input options, > rather than allowing
freeform input.
```

 * As for further guidance: There's currently only a _Wiki_ article that is an unpolished collection of background information, scenarios to avoid, and also legitimate uses: https://github.com/MicrosoftDocs/PowerShell-Docs/wiki/The-case-for-and-against-Invoke-Expression
     * Presumably because of its unpolished state, it isn't linked to from the `Invoke-Expression` help topic, which makes it hard to discover.

**Comment by Jaykul** [comment]
You asked for reasons why we use it:

## 1. Testability.

Show me a way to **mock** the & call operator in Pester tests, and I could remove the vast majority of `Invoke-Expression` from my production code.

## 2. Testing.

In our tests (and, incidentally, in the tests for Microsoft's DSC), there are frequent uses of Invoke-Expression to dynamically generate larger suites of tests without copy/pasting ... 

## 3. Importing metadata that aren't module manifests

People who are not using my `Configuration` module or something like it, frequently use `Invoke-Expression (Get-Content...)` instead of Import-LocalizedData because:
1. Import-LocalizedData is overly complicated to use for miscellaneous files because it has no -Path parameter
2. Import-LocalizedData was too restrictive on the commands it considered "safe" in metadata files (not even allowing all the same ones as module manifests). 
3. `Parser.ParseInput` or `Parser.ParseFile` combined with `ScriptBlock.CheckRestrictedLanguage` are complicated to use (especially since they've changed across PS versions)

## 4. Invoking code someone else gives us.

Yeah, this obviously includes a lot of unsafe things like:
```PowerShell
Invoke-Expression (Invoke-RestMethod 'https://chocolatey.org/install.ps1')
```

But it also includes _really valid_ things like loading tab-completers from third party utilities:
```PowerShell
conda "shell.powershell" "hook" | Out-String | Invoke-Expression
```

## My Conclusion

Obviously a lot of people will be very upset (as we always are when you break things) if you make our existing production code start producing warnings just because you've decided something you wrote a decade ago was a bad idea.

ScriptAnalyzer rules are a more polite way to coerce people to do things differently

This item has been open for way too long. There needs to be an automatic time-out on issues opened for the purpose of discussion.

**Comment by sdwheeler** [comment]
@Jaykul Thanks, these are good examples.

**Comment by mklement0** [comment]
@sdwheeler, probably worth adding to https://github.com/MicrosoftDocs/PowerShell-Docs/wiki/The-case-for-and-against-Invoke-Expression.

@Jaykul, an aside re the `Invoke-Expression` /  `Invoke-RestMethod` combo for executing downloaded script contents: the need for that would go away:

*  if we had support for `Invoke-WebRequest` output in `Invoke-Command` (e.g.,
`iwr https://aka.ms/install-powershell.ps1 | icm`), as proposed in https://github.com/PowerShell/PowerShell/issues/5909#issuecomment-46119220.

* or, as an out-of-process alternative, if `pwsh -` worked properly with code piped to it (e.g.,
`irm https://aka.ms/install-powershell.ps1 | pwsh`)  - see #3223.

**Comment by vexx32** [comment]
For point #1, you could just wrap the code that invokes via `&` in a function and mock that. But yes, it's less convenient, for sure.

I wonder if it will be really needed for #2 with Pester's new -ForEach functionality to generate tests a bit more dynamically, but yeah there are definitely some cases where things are much neater with code gen than other methods.

For #3 I think we may have other issues around that, but I agree it would be _very_ nice to make it easier to import files like that natively without invoke-expression.

And heck, half the time I won't even really bother with Invoke-Expression for downloaded scripts like that, I'll just construct a scriptblock and use `&` to invoke it.

---

But yeah, I don't think deprecating it is really the right move. There are better alternatives for _some_ of its use cases, but not all, at least at the moment.

**Comment by sdwheeler** [comment]
I plan to incorporate the feedback from here and move the wiki article to the 'Deep Dives' section of the docs.

**Comment by rjmholt** [comment]
> Show me a way to mock the & call operator in Pester tests, and I could remove the vast majority of Invoke-Expression from my production code.

Could you expand on the scenario here, or give an example of the mocking for a test (i.e. where the mocking steps in for testing)?

**Comment by StewartRyanSony** [comment]
@Jaykul touched on this, but I don't know of another way to setup tab completion of a parameter there the possible values exist outside of the script. Here's an example.

```powershell
param(
    [Parameter(Mandatory=$false, Position=0, ParameterSetName='Environment')]
    [ArgumentCompleter(
        {
            param ($Command, $Parameter, $WordToComplete, $CommandAst, $FakeBoundParams)
            $ParamPath = "$($PSScriptRoot ? $PSScriptRoot : $(Get-Location))\Environments.json"
            if (!(Test-Path $ParamPath)) {"Environments.json file not found. Please be sure this file is in the same directory as this script. Otherwise, please use the other parameters"}
            else {
                Get-Content $ParamPath | ConvertFrom-Json | Select-Object -ExpandProperty Name | Where-Object {$_ -like "$WordToComplete*"}
            }
        }
    )]
    [ValidateScript(
        {
            $_ -in (Invoke-Command {
                $ParamPath = "$($PSScriptRoot ? $PSScriptRoot : $(Get-Location))\Environments.json"
                Get-Content $ParamPath | ConvertFrom-Json | Select-Object -ExpandProperty Name
            })
        }
    )]
    [Alias('Environment')]
    [String]$Environment = "Dev",
...```

**Comment by rjmholt** [comment]
> Here's an example.

I can't see `Invoke-Expression` used in the example. You could probably also remove `Invoke-Command` too:

```powershell
[ValidateScript(
    {
        $valuesPath = Join-Path ($PSScriptRoot ? $PSScriptRoot : Get-Location) 'Environments.json'
        $validValues = Get-Content -Raw -LiteralPath $valuesPath | ConvertFrom-Json | Select-Object -ExpandProperty Name
        return $_ -in $validValues
    }
)]
```

**Comment by StewartRyanSony** [comment]
oh, you're right. I used Invoke-Command. I think I changed that and forgot about it. Never mind

**Comment by aescolastico** [comment]
Is using '&' a suitable replacement for running exes? I use Start-Process sometimes, but that doesn't cleanly handle sdtout, sdtin and sdterr streams

**Comment by yecril71pl** [comment]
> Is using '&' a suitable replacement for running exes, or are they the same thing? I use Start-Process sometimes, but that doesn't cleanly handle sdtout, sdtin and sdterr streams

No and no.

**Comment by mklement0** [comment]
@aescolastico: The `&` operator does _not_ evaluate PowerShell source code stored in a string the way that `Invoke-Expression` does. 

It only accepts a _command name or path_ as its operand (or, alternatively a script block), so for running external programs the arguments must still be passed separately. 

In fact, the only reason `&` is ever needed when calling external programs a _syntactic_ one: So that PowerShell knows you're executing a command, `&` must be used if the executable file path is specified in quotes and/or contains variable references / subexpression (`& $HOME\bin\foo.exe bar`); otherwise, just invoke the executable directly (`C:\Users\jdoe\bin\foo.exe bar`).

If you need to construct arguments programmatically, it is much cleaner to create them as elements of an array first, which you can then pass as a whole (`$arguments = 'bar', 'baz'; foo.exe $arguments`) - beware the issues around passing empty-string arguments and arguments with embedded `"` chars., however - see #1995.

`Start-Process`, which allows you to pass the list of arguments for an external executable as a single string, is the wrong tool for synchronously invoking external programs that are console applications - as you've observed, you lose integration with PowerShell's streams.

**Comment by yecril71pl** [comment]
> In fact, the only reason `&` is ever needed when calling external programs is a _syntactic_ one: So that PowerShell knows you're executing a command, `&` must be used if the executable file path ~is specified in quotes and/or~ contains variable references / subexpressions (`& "C:\Program Files\Acme\foo.exe" bar` or `& $HOME\bin\foo.exe bar`); otherwise, just invoke the executable directly (`C:\Users\jdoe\bin\foo.exe bar`).

`C:\"Program Files\Acme\foo.exe" bar` should work too.

**Comment by mklement0** [comment]
Yes, what matters is whether the _first character_ is a quote - because that determines the [parsing mode](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Parsing) - so technically the rule is "specified in quotes _as a whole_ or _starts with a quoted token_ and/or contains variable references / subexpressions".

Pragmatically speaking, since `&` _always_ works, the simplified rule "specified in quotes and/or contains variable references / subexpressions" is easy to remember and should suffice.

**Comment by yecril71pl** [comment]
> @sdwheeler, probably worth adding to https://github.com/MicrosoftDocs/PowerShell-Docs/wiki/The-case-for-and-against-Invoke-Expression.

####  The-case-for-and-against-Invoke-Expression.md 
```patch
index e01a666..44f0936 100644
@@ -49,7 +49,7 @@ There are probably a few layers of "I want to use Invoke-Expression":
     - To [parse a string with embedded quoting](https://stackoverflow.com/a/51229037/45375), _with extra precautions_
     - To [parse command lines stored in a file](https://stackoverflow.com/a/60328737/45375), _if trusted_
 
-Despite the legitimacy of some use cases, I think they're as rare as hens' teeth and we shouldn't
+Despite the legitimacy of some use cases, I think they're as rare as hen&rsquo;s teeth and we shouldn't
 list them; dynamically creating classes or executing output from a command are, even without
 `Invoke-Expression`, things users should only be doing as a last resort.
```

**Comment by aescolastico** [comment]
> @aescolastico: The `&` operator does _not_ evaluate PowerShell source code stored in a string the way that `Invoke-Expression` does.
> 
> It only accepts a _command name or path_ as its operand (or, alternatively a script block), so for running external programs the arguments must still be passed separately.
> 
> In fact, the only reason `&` is ever needed when calling external programs is a _syntactic_ one: So that PowerShell knows you're executing a command, `&` must be used if the executable file path is specified in quotes and/or contains variable references / subexpressions (`& "C:\Program Files\Acme\foo.exe" bar` or `& $HOME\bin\foo.exe bar`); otherwise, just invoke the executable directly (`C:\Users\jdoe\bin\foo.exe bar`).
> 
> If you need to construct arguments programmatically, it is much cleaner to create them as elements of an array first, which you can then pass as a whole (`$arguments = 'bar', 'baz'; foo.exe $arguments`) - beware the issues around passing empty-string arguments and arguments with embedded `"` chars., however - see #1995.
> 
> `Start-Process`, which allows you to pass the list of arguments for an external executable as a single string, is the wrong tool for synchronously invoking external programs that are console applications - as you've observed, you lose integration with PowerShell's streams, even if you run the process synchronously (`-Wait`) in the same console window (`-NoNewWindow`, implied on Unix-like platforms).

This is a very clear and succinct explanation. Thank you.

Do you know if there are ways of invoking scripts hosted online without using iex?
ie 
DownloadString or BITs $scriptURL | Invoke-Expression

I guess you could download the script to a temp ps1 file, call that, then delete it. But I'm curious if there's a way to do it entirely in memory.

**Comment by yecril71pl** [comment]
> I guess you could download the script to a temp ps1 file, call that, then delete it. But I'm curious if there's a way to do it entirely in memory.

Executing a script saved to a temporary file does it entirely in memory.

**Comment by Jaykul** [comment]
@mklement0 wrote: 
>The `&` operator does _not_ evaluate PowerShell source code stored in a string the way that `Invoke-Expression` does.
> 
> It only accepts a _command name or path_ as its operand (or, alternatively a script block), so for running external programs the arguments must still be passed separately.

Or any of the command *Info objects, or a module object and _then_ a scriptblock, or  ... well, you get the drift, there's actually a lot of things, but in all cases, the thing `&` applies to an an object treated literally, not parsed...

@yecril71pl writes:
> Executing a script saved to a temporary file does it entirely in memory.

Huh? Are we being super pedantic here for a reason? 
Yes, it _executes_ in memory, _because that's what execution means_.
But the point he's making is that _downloading it to file_ creates a file _on disk_ which needs cleaning up ... 

@mklement0 also wrote, much earlier: 
> the need for this idiom would go away:
> 
>     * if we had support for `Invoke-WebRequest` output in `Invoke-Command` (e.g.,
>       `iwr https://aka.ms/install-powershell.ps1 | icm`), as proposed in [#5909 (comment)](https://github.com/PowerShell/PowerShell/issues/5909#issuecomment-461192202), and things would get even simpler if #8835 gets implemented.
>     * or, as an out-of-process alternative, if `pwsh -` worked properly with code piped to it (e.g.,
>       `irm https://aka.ms/install-powershell.ps1 | pwsh`)  - see #3223.

Yes, obviously if we did either of those things, we would certainly remove the need for `Invoke-Expression` -- but what would be the point? If we did that, we would have simply transferred all the "inherent security risks" the OP reffered to into new places which would be **much** harder to flag in a code review or static analyzer.

**Comment by mklement0** [comment]
> Yes, obviously if we did either of those things, we would certainly remove the need for `Invoke-Expression`

The primary reason for not using `Invoke-Expression` (`iex`) is that is is _inadequate_ for running the content of script files directly from the web:

* The code runs directly in the caller's scope - a pollution of that scope that is likely unwanted.
* If the downloaded code contains an `exit` statement, the session exits as a whole.
* (There are also 2 bugs, though they obviously could and should be fixed:  #8778 and #8815)

Rather than extend `Invoke-Expression`, it makes more sense to enhance `Invoke-Command`, [as proposed](https://github.com/PowerShell/PowerShell/issues/5909#issuecomment-461192202).

* `Invoke-Command` (`icm`) runs code in a child scope by default, while allowing running directly in the caller's scope as an opt-in via `-NoNewScope`

* With an `Invoke-WebRequest` (`iwr`) output object as input, it would have to treat `exit` as simply exiting the child scope, i.e. it would have to treat the code as if it were a script _file_.

```powershell
# WISHFUL THINKING
iwr https://aka.ms/install-powershell.ps1 | icm -- -Preview"
```

---

Also, to recap @SteveL-MSFT's workaround for use of `Invoke-Expression`, which solves the _scope_ issue, but not the `exit` issue:

```powershell
iex "& { $(irm https://aka.ms/install-powershell.ps1) } -Preview"
```

**Comment by mklement0** [comment]
I forgot to spell out another important limitation of `Invoke-Expression` with respect to direct execution of an online script: the inability to pass _arguments_ (which, again, `Invoke-Command` already supports); I've amended the previous comment.

**Comment by yecril71pl** [comment]
> > Executing a script saved to a temporary file does it entirely in memory.
> 
> Huh? Are we being super pedantic here for a reason?
> Yes, it _executes_ in memory, _because that's what execution means_.
> But the point he's making is that _downloading it to file_ creates a file _on disk_ which needs cleaning up ...

The storage for (small) temporary files is in RAM.

**Comment by JamesWTruher** [comment]
@PowerShell/powershell-committee reviewed this 5/11/2022
We definitely do want to deprecate `Invoke-Expression` - when you _need_ it, you really _need_ it. We also have _no_ plans to remove it from the product, so marking it as deprecated which implies that it won't always be there is misleading. However, we do agree that we could possibly provide more guidance and information as to why it's such a dangerous tool.

**Comment by sdwheeler** [comment]
I opened this issue to have a discussion about the idea. This really should have been a [Discussion](https://github.com/PowerShell/PowerShell/discussions) thread, but **Discussions** were not enabled at the time.

The question has been adequately answered. I don't think we should deprecate `Invoke-Expression`. However, we should probably have better guideance for the risks involved with using it. The article [Preventing script injection attacks](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/security/preventing-script-injection) is a start.
